### PR TITLE
Fix #18868, MS Basic trill playback incorrect when spanning multiple notes

### LINKED ISSUE #18868: Tied notes and trill's major/minor doesn't work properly (Ms Basic)
### Issue type

General playback bug

### Bug description

When using MS Basic sounds, if the trill note lengthens with a tie to the other measure, the trill's major/minor ' feature does not work properly.

### Steps to reproduce

if you can watch the video i tried to show it there

### Screenshots/Screen recordings

https://youtu.be/vxxezOi3RKk

### MuseScore Version

MuseScore 4.1.1

### Regression

I don't know

### Operating system

Linux Mint 21, Windows 10

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wizofaus**
Resolves: #18868

- [X] I signed the [CLA](https://musescore.org/en/cla)
- [X] The title of the PR describes the problem it addresses
- [X] Each commit's message describes its purpose and effects, and references the issue it resolves
- [X] If changes are extensive, there is a sequence of easily reviewable commits
- [X] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [X] There are no unnecessary changes
- [X] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by wizofaus** [comment]
Assumption is this is that regardless of tie, if a trill spanner covers a note, the properties of that trill element should be used to determine playback (existing behaviour failed to find the trill element and just used the default playback settings).

File: src/engraving/dom/chord.cpp [hunk 1]
```diff
@@ -3231,6 +3231,11 @@ Ornament* Chord::findOrnament() const
             return toTrill(spanner)->ornament();
         }
     }
+    for (Spanner* spanner : m_endingSpanners) {
+        if (spanner->isTrill()) {
+            return toTrill(spanner)->ornament();
+        }
+    }
```
Comment (wizofaus) [inline src/engraving/dom/chord.cpp:3238]: this even seems to work for trill spanners that start on previous note and extend to next note, surprisingly.

**Review by RomanPudashkin — APPROVED** [review]

File: src/engraving/dom/chord.cpp (hunk 1 above)
Comment (wizofaus) [reply src/engraving/dom/chord.cpp:3238]: Please attach your score then, this definitely fixed it for me

Comment (orkuun) [reply src/engraving/dom/chord.cpp:3238]: I couldn't find/know how to add a file with *.mscz extension here. I am sending you as a screenshot. But in the meantime I discovered something. When I save the file and open it again, it does not make any errors, but when I change the trill/properties/ornament/interval settings, it does not play correctly.
screens

**Comment by zacjansheski** [comment]
I'm not getting the desired results initially, but if I pitch up/down (update playback) I get the correct playback.

There is an issue with reverting back to the default trill (see end)

Also looks like there is an engraving issue (horizontal spacing) I need to look into

https://github.com/musescore/MuseScore/assets/69917893/ad56dc13-e93e-46b2-85d5-79da7a27623c

File: src/engraving/dom/chord.cpp (hunk 1 above)
Comment (wizofaus) [reply src/engraving/dom/chord.cpp:3238]: Just rename the .mscz to .zip (or anything really).

**Comment by wizofaus** [comment]
Ok resolved that, it needs special case logic after editing an ornament to determine if there are any trill spanners attached to the corresponding note to ensure the "end tick" for the playback update range is set correctly.

File: src/engraving/dom/chord.cpp (hunk 1 above)
Comment (wizofaus) [reply src/engraving/dom/chord.cpp:3238]: Anyway I've pushed a separate fix that ensures after editing a trill the playback for the whole length of the trill is updated (vs just that for the note its attached to).

**Comment by cbjeukendrup** [comment]
> resolved that

Did you intend to push these changes? (I don't see new changes, namely)

**Comment by wizofaus** [comment]
Oh, weird, could swear I'd pushed - just a sec.

**Review by RomanPudashkin — APPROVED** [review]

**Comment by zacjansheski** [comment]
I'm hitting an interesting crash on this that is not in master:

1. Make a score with a flute and a Bb Clarinet (MS Basic or MuseSounds, doesn't matter)
2. Write a two measure concert B on both
3. Add a trill to both
4. With both selected, change the trill interval to a diminished 7th


https://github.com/musescore/MuseScore/assets/69917893/6fbdc650-1ddd-4930-9b5d-66cdef56f7de

**Comment by wizofaus** [comment]
Thanks, investigating. Did you confirm the bug as reported was fixed otherwise?

**Comment by zacjansheski** [comment]
@wizofaus I can't reproduce in master if that is what you are asking

**Comment by wizofaus** [comment]
No, I'm asking if when testing with this branch the unexpected behaviour as reported with trill playback has been fixed properly?

**Comment by zacjansheski** [comment]
It's working much better, but not completely fixed.  There is still inconsistency with the trill when undoing

https://github.com/musescore/MuseScore/assets/69917893/40926340-79f3-4840-9464-1fdba3d0502f

**Comment by wizofaus** [comment]
Ah ok, yep, that figures.

**Comment by wizofaus** [comment]
I will say that crash doesn't appear to have anything to do with what I changed (it's while computing the layout), not sure why it wouldn't happen with master.

**Comment by wizofaus** [comment]
Ok, fixed the crash, but the really odd thing is that after doing so I can't reproduce the undo problem, even though I assumed that would be unrelated!

**Comment by wizofaus** [comment]
Ok pushed two more fixes:

a) prevent crash on undo after changing a multi-note trill to use an interval over a second (where an auxiliary note is shown)
b) ensure that undo of changing interval for a multi-note trill correctly updates playback for full length of trill

(multi-note trill is any trill spanning more than one note, typically should be the same note, doesn't need to be tied)

File: src/engraving/dom/chord.cpp (hunk 1 above)
Comment (wizofaus) [reply src/engraving/dom/chord.cpp:3240]: Turns out it's important NOT to return the ornament if it's being queried for layout purposes (particularly with regard to auxiliary notes to indicate trill interval).

File: src/engraving/playback/playbackmodel.cpp [hunk 2]
```diff
@@ -892,6 +892,13 @@ PlaybackModel::TickBoundaries PlaybackModel::tickBoundaries(const ScoreChangesRa
             result.tickFrom = std::min(result.tickFrom, firstTiedNote->tick().ticks());
             result.tickTo = std::max(result.tickTo, lastTiedNote->tick().ticks());
         }
+        if (item->parent() && item->parent()->isChord()) {
```
Comment (wizofaus) [inline src/engraving/playback/playbackmodel.cpp:895]: Originally I had an "else" in here, but that didn't work in the case of "undo", where the ornament object being modified doesn't have a valid parent at this point.  Technically this code could probably be moved inside the "`if (item->isNote()) {` block above, but it's safer enough here, given the only purpose is to ensure that the end-tick of the playback update region covers the longest possible trill spanner that might be relevant.

**Comment by zacjansheski** [comment]
Tested on MacOS 14, Windows 11, Ubuntu 22.04.3. Approved
https://github.com/musescore/MuseScore/issues/18868 FIXED

File: src/engraving/dom/chord.cpp [hunk 3]
```diff
@@ -3219,7 +3219,7 @@ void Chord::computeKerningExceptions()
     }
 }
 
-Ornament* Chord::findOrnament() const
+Ornament* Chord::findOrnament(bool forPlayback) const
```
Comment (RomanPudashkin) [inline src/engraving/dom/chord.cpp:3222]: Could you rename this flag? Something like "includingEndingSpanners" (or something else)

Comment (wizofaus) [reply src/engraving/dom/chord.cpp:3222]: My intention was that the name indicates the purpose for calling the function, i.e. whether you're doing it for layout or playback purposes - that currently this means checking "ending spanners" when determining playback is an implementation detail that might change in future.

Comment (RomanPudashkin) [reply src/engraving/dom/chord.cpp:3222]: Functions should never know by whom or how they will be called

**Pull request merged**